lib/charms/opensearch/v0/opensearch_distro.py

```diff
@@ -431,15 +433,31 @@ def current(self) -> Node:
                 temperature=current_node.get("attributes", {}).get("temp"),
             )
         except OpenSearchHttpError:
-            # we try to get the most accurate description of the node
-            conf_on_disk = self.config.load("opensearch.yml")
+            # we try to get the most accurate description of the node from the static config
+            conf = self.config.load("opensearch.yml")
+            deployment_desc = self._charm.opensearch_peer_cm.deployment_desc()
+            try:
+                roles = conf["node.roles"]
+            except KeyError:
+                if deployment_desc.start == StartMode.WITH_PROVIDED_ROLES:
+                    roles = deployment_desc.config.roles
+                else:
+                    roles = GeneratedRoles
+
+            try:
+                temperature = conf["node.attr.temp"]
+            except KeyError:
+                temperature = deployment_desc.config.data_temperature
+            if not roles:
+                raise OpenSearchError("Can not determine app name and/or roles.")
+
             return Node(
                 name=self._charm.unit_name,
-                roles=conf_on_disk["node.roles"],
+                roles=roles,
                 ip=self._charm.unit_ip,
-                app=App(id=conf_on_disk.get("node.attr.app_id")),
+                app=deployment_desc.app,
```

Comment:
Questoin @Mehdi-Bendriss @reneradoi @phvalguima 

Normally `OpenSearchPeerClustersManager.deployment_desc()` returns a data structure with values from the databag -- or `None` if databag information couldn't be retrieved.

I.e. the `deployment_desc` variable techinically can be `None`, resulting in the above snippet failing at muliple levels.

Now, the original bug https://github.com/canonical/opensearch-operator/issues/222 was raised at unit shutdown.

Is it possible that `OpenSearchDistribution.current()` may be executed at a point when peer data may not be available anymore? (Why?)

Reply:
In other words, are we sure that this solution is not duplicating the already existing new issue https://github.com/canonical/opensearch-operator/issues/422 (I just opened about a week ago), throwing precisely the error that I foresee will be also happening if we merge the code above:

```
  File "/var/lib/juju/agents/unit-opensearch-25/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 1573, in unit_name
    return format_unit_name(self.unit, app=self.opensearch_peer_cm.deployment_desc().app)
AttributeError: 'NoneType' object has no attribute 'app'
```

...?

Reply:
@Mehdi-Bendriss  earlier [suggested](https://github.com/canonical/opensearch-operator/pull/412#discussion_r1729248248) to follow this strategy for `roles`: disk, peer relation, error if neither.

Now, the `app_id` is stored in the local config (see screenshot below), and we _did_ check it (only) there with the original code. 

I'd like to understand why do we replace the original strategy, instead of extending it with an additional check (as done for `roles`)?

Also I'm expressing that I'm not sure if the current proposal is solid but I may be mistaken.

![image](https://github.com/user-attachments/assets/8b02b31a-6b93-49b9-a411-da948e1bca49)

Reply:
Summarizing live conversation: reasons to use different sources for `app` vs. `roles` got clarified (see upcoming comments around related code).

It also seemed reasonable to fall back to the original (local config) for `app` in case the Peer Relation may not be available (as in corner cases like the error above).